File: 234-palindrome-linked-list/234-palindrome-linked-list.c
#define MAXSIZE 100000

typedef struct s{
    int val[MAXSIZE];
    int top;
}Stack;

Stack *createStack(){
    Stack *stk = malloc(sizeof(Stack));
    stk->top = 0;
    return stk;
}

void push(Stack* stk, int new_val){
    if(stk->top == MAXSIZE)
    {
        return;
    }
    else{
        stk->val[stk->top++] = new_val;
    }
}

int pop(Stack* stk){
    if(stk->top == 0){
        return;
    }
    else{
        return stk->val[--stk->top];
    }
}

bool isPalindrome(struct ListNode* head){

    struct ListNode* slow = head;
    if(slow->next == NULL){
        return true;
    }
    
    struct ListNode* fast = head;
    Stack* stk;
    stk = createStack();
    
    while(fast && fast->next){

        push(stk, slow->val);
        slow = slow->next;
        fast = fast->next->next;
    }
    if(fast){
        slow = slow->next;
    }
    while(slow){
        int tmp = pop(stk);
        if(tmp!=slow->val){
            return false;
        }
        slow= slow->next;
    }
    return true;
}

File: 234-palindrome-linked-list/234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode tmp = head;
        int len = 0;
        while(tmp != null){
            len++;
            tmp = tmp.next;
        }
        Stack<Integer> st = new Stack();
        int flag = len % 2;
        
        for(int i = 0; i < len; i++){
            if(i == len/2 && flag == 1){
                head = head.next;
                continue;
            }
            if(i >= len/2){
                if(!st.isEmpty()){
                    if(st.peek() != head.val)
                        return false;
                    st.pop();
                }
            }
            else
                st.push(head.val);
            head = head.next;
        }
        return true;
    }
}